Codes/Clean Data/getHasbrouck.R

```diff
@@ -0,0 +1,108 @@
+function(indexExe, lags, price, direction, sz) {
+  #vector autoregressive (VAR) system of differences in log prices ($r_t=\Delta p_t$) and trade-related characteristics $x_t$:
+  
+  y <- diff(log(price[indexExe]))
+  
+  #The column vector of trade variables $\mathbf{x}_t$ includes:
+  #(1) a sign indicator reflecting the direction of the trade,
+  #(2) signed trading volume, and
+  #(3) the signed square root of trading volume.
+  
+  x1 <-
+    direction[indexExe] * (-1)
+  x1 <-
+    x1[-1] #note that direction needs to be multiplied by -1, as it refers to sign on submission (not execution)
+  x2 <- sz[indexExe]
+  x2 <- x2[-1] * x1
+  x3 <- sz[indexExe]
+  x3 <- sqrt(x3[-1]) * x1
+  
+  ##step one: get residuals from VAR
+  model <- cbind(y, x1, x2, x3)
+  dimx <- ncol(model)
+  test <- try(VAR(model,
+                  p = lags,
+                  output = F,
+                  include.mean = F), silent = T)
+  if (class(test) == "try-error") {
+    return(c(NA, NA, NA))
+    break

```

Comment:
You are correct, the "return" here is indeed redundant.

Reply:
Are you sure return is redundant? I thought "break" is redundant. I know the "break" will break the loop of the script that is calling this function. But please be careful, because "break" without the loop inside the same function makes the function  unsafe for reusing in other script, when there is no loop the script will crash!

Reply:
Indeed you delete break! Cool. Thank you!